frame/composable-traits/src/staking/lock.rs

```diff
@@ -0,0 +1,39 @@
+use crate::time::{DurationSeconds, Timestamp};
+use codec::{Decode, Encode};
+use frame_support::dispatch::DispatchResult;
+
+use core::fmt::Debug;
+
+use scale_info::TypeInfo;
+use sp_runtime::Perbill;
+
+/// defines staking duration, rewards and early unstake penalty
+#[derive(Debug, PartialEq, Eq, Copy, Clone, Encode, Decode, TypeInfo)]
+pub struct LockConfig<DurationPresets> {
+	/// The possible locking duration.
+	pub duration_presets: DurationPresets,
+	/// The penalty applied if a staker unstake before the end date.
+	/// In case of zero penalty, you cannot unlock before it duration ends.
+	pub unlock_penalty: Perbill,
```

Comment:
instead of checking against zero, make this an enum `Unlock = Penalty Perbill | NotUnlockable` (or something similar)

Reply:
that will not work, until `UnlockPenalty = NonZeroPerbill | NotUnlockable`. Other point, is that if unlock penalty is 0.0000001 ? So it will take all amount anyway.

Reply:
NonZeroPerbill - will try on vacation to get macro. I have example, so not macroed yet.